dask/layers.py

```diff
@@ -1239,3 +1262,341 @@ def __repr__(self):
+        while parts > 1:
+            parts = math.ceil(parts / self.split_every)
+            self.widths.append(parts)
+        self.height = len(self.widths)
+
+        # Set default output_splits
+        if self.output_splits is None:
+            self.output_splits = list(range(self.split_out or 1))
+
+    def _check_key(self, *task):
+        # Helper function to remove the last
+        # (split_out) element in a task key
+        # when bool(split_out) is False
+        return task if self.split_out else task[:-1]
+
+    def _define_task(self, node_list, output_node=False):
+        # Define nested concatenation and func task
+        conc = (self.concat_func, node_list)
+        if output_node and self.finalize_func:
+            func = self.finalize_func
+            kwargs = self.finalize_kwargs
+        else:
+            func = self.tree_node_func
+            kwargs = self.tree_node_kwargs
+        if kwargs:
+            return (apply, func, [conc], kwargs)
+        else:
+            return (func, conc)
+
+    def _construct_graph(self, on_scheduler=False):
+        """Construct graph for a tree reduction."""
+
+        dsk = {}
+        if not self.output_splits:
+            return dsk
+
+        # Deal with pickled user-defined functions
+        if on_scheduler:
+            self.tree_node_func = CallablePickled(self.tree_node_func)
+            self.concat_func = CallablePickled(self.concat_func)
+            if self.finalize_func:
+                self.finalize_func = CallablePickled(self.finalize_func)
+
+        # Deal with `bool(split_out) == True`.
+        # These cases require that the input tasks
+        # ruturn a type that enables getitem operation
+        # with indices: [0, split_out)
+        # Therefore, we must add "getitem" tasks to
+        # select the appropriate element for each split
+        name_input_use = self.name_input
+        if self.split_out:
+            name_input_use += "-split"
+            for s in self.output_splits:
+                for p in range(self.npartitions_input):
+                    dsk[(name_input_use, p, s)] = (
+                        operator.getitem,
+                        (self.name_input, p),
+                        s,
+                    )
+
+        if self.height >= 2:
+            # Loop over output splits
+            for s in self.output_splits:
+                # Loop over reduction levels
+                for depth in range(1, self.height):
+                    # Loop over reduction groups
+                    for group in range(self.widths[depth]):
+                        # Calculate inputs for the current group
+                        p_max = self.widths[depth - 1]
+                        lstart = self.split_every * group
+                        lstop = min(lstart + self.split_every, p_max)
+                        if depth == 1:
+                            # Input nodes are from input layer
+                            node_list = [
```

Comment:
```suggestion
                            input_keys = [
```
Easier to understand what this is